org.osgi.util.promise/src/org/osgi/util/promise/Promise.java

```diff
@@ -503,4 +503,166 @@
+	 *            type. Must not be {@code null}.
+	 * @param failureType The type of failure for which this callback will be
+	 *            called. If the failure is not an instance of the specified
+	 *            failure type, the specified callback will not be called. Must
+	 *            not be {@code null}.
+	 * @return This Promise.
+	 * @since 1.3
+	 */
+	Promise<T> onFailure(Consumer< ? super Throwable> failure,
+			Class< ? > failureType);
+
+	/**
+	 * Recover from a failure of this Promise with a recovery value if the
+	 * failure is an instance of a failure type.
+	 * <p>
+	 * If this Promise is successfully resolved, the returned Promise must be
+	 * resolved with the value of this Promise.
+	 * <p>
+	 * If this Promise is resolved with a failure and the failure is not an
+	 * instance of the specified failure type, the returned Promise must be
+	 * failed with the failure of this Promise.
+	 * <p>
+	 * If this Promise is resolved with a failure and the failure is an instance
+	 * of the specified failure type, the specified Function is applied to this
+	 * Promise to produce a recovery value.
+	 * <ul>
+	 * <li>If the recovery value is not {@code null}, the returned Promise must
+	 * be resolved with the recovery value.</li>
+	 * <li>If the recovery value is {@code null}, the returned Promise must be
+	 * failed with the failure of this Promise.</li>
+	 * <li>If the specified Function throws an exception, the returned Promise
+	 * must be failed with that exception.</li>
+	 * </ul>
+	 * <p>
+	 * To recover from a failure of this Promise with a recovery value of
+	 * {@code null}, the {@link #recoverWith(Function, Class)} method must be
+	 * used. The specified Function for {@link #recoverWith(Function, Class)}
+	 * can return {@code Promises.resolved(null)} to supply the desired
+	 * {@code null} value.
+	 * <p>
+	 * This method may be called at any time including before and after this
+	 * Promise has been resolved.
+	 * 
+	 * @param recovery If this Promise resolves with a failure and the failure
+	 *            is an instance of the specified failure type, the specified
+	 *            Function is called to produce a recovery value to be used to
+	 *            resolve the returned Promise. Must not be {@code null}.
+	 * @param failureType The type of failure for which the specified recovery
+	 *            will be used. If the failure is not an instance of the failure
+	 *            type, the specified recovery will not be called and the
+	 *            returned Promise must be failed with the failure of this
+	 *            Promise. Must not be {@code null}.
+	 * @return A Promise that resolves with the value of this Promise or
+	 *         recovers from the failure of this Promise if the failure is an
+	 *         instance of the specified failure type.
+	 * @since 1.3
+	 */
+	Promise<T> recover(Function<Promise< ? >, ? extends T> recovery,
+			Class< ? > failureType);
```

Comment:
Constraining the failureType's type to `Class< ? extends Throwable >` would prevent non-exception types to be passed as failure type.